[FLANG][OpenMP]Add frontend support for ASSUME and ASSUMES

Leporacanthicus opened the pull request:
Enough suport to parse correctly formed directives of !$OMP ASSUME and !$OMP ASSUMES with teh related clauses that go with them: ABSENT, CONTAINS, NO_OPENPP, NO_OPENMP_ROUTINES, NO_PARALLELISM and HOLDS.

Tests added for unparsing and dump parse-tree.

Semantics support is very minimal and no specific tests added.

The lowering will hit a TODO, and there are tests in Lower/OpenMP/Todo to make it clear that this is currently expected behaviour.

Leporacanthicus committed:
[FLANG][OpenMP]Add frontend support for ASSUME and ASSUMES

Enough suport to parse correctly formed directives of !$OMP ASSUME
and !$OMP ASSUMES with teh related clauses that go with them:
ABSENT, CONTAINS, NO_OPENPP, NO_OPENMP_ROUTINES, NO_PARALLELISM
and HOLDS.

Tests added for unparsing and dump parse-tree.

Semantics support is very minimal and no specific tests added.

The lowering will hit a TODO, and there are tests in Lower/OpenMP/Todo
to make it clear that this is currently expected behaviour.

kiranchandramohan commented:
Assumes directive cannot occur in the execution part of Fortran. See the following in Section 8.3.2 of OpenMP 5.2 standard. So it is not an executable directive.
```
The assumes directive may only appear in the specification part of a module or subprogram after any USE statement, any IMPORT statement, and any IMPLICIT statement.
```

Leporacanthicus committed:
Review comments

Leporacanthicus commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3971,6 +3992,15 @@ struct OmpGrainsizeClause {
+struct OmpHoldsClause {
+  using ExprList = std::list<common::Indirection<Expr>>;
+  WRAPPER_CLASS_BOILERPLATE(OmpHoldsClause, ExprList);
+};
```
Nope, me being a bit enthusiastic about what that should do! :)

Leporacanthicus commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -4174,6 +4219,44 @@ struct OmpClauseList {
+
+struct OmpEndAssumesDirective {
+  TUPLE_CLASS_BOILERPLATE(OmpEndAssumesDirective);
+  std::tuple<Verbatim> t;
```
Fixed.

Leporacanthicus commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -3740,6 +3740,19 @@ struct OmpVariableCategory {
+struct OmpDirectiveNameEntry {
+  WRAPPER_CLASS_BOILERPLATE(OmpDirectiveNameEntry, llvm::omp::Directive);
+};
+using OmpDirectiveList = std::list<OmpDirectiveNameEntry>;
```
Fixed.

Leporacanthicus commented:
> Assumes directive cannot occur in the execution part of Fortran. See the following in Section 8.3.2 of OpenMP 5.2 standard. So it is not an executable directive.
> 
> ```
> The assumes directive may only appear in the specification part of a module or subprogram after any USE statement, any IMPORT statement, and any IMPLICIT statement.
> ```

Thanks for pointing this out, I completely missed that this was the distinction between BEGIN ASSUMES and ASSUMES - I thought it was a bit weird that there was two different ways to express something very similar. :) 

Now fixed.

Leporacanthicus committed:
Remove incorrect "begin" keyword.

This is only used in C++, so do not use it in the Fortran compiler.

Updated tests to suit.

kparzysz commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -4174,6 +4215,49 @@ struct OmpClauseList {
+  CharBlock source;
+};
+
+struct OpenMPAssumesConstruct {
+  TUPLE_CLASS_BOILERPLATE(OpenMPAssumesConstruct);
+  std::tuple<OmpAssumesDirective, OmpAssumesPartConstruct,
+      OmpEndAssumesDirective>
+      t;
+  CharBlock source;
```
This should be `Assume` (note no `s` at the end).  The `ASSUME` directive is associated with a block (and it has `END ASSUME` counterpart).

kiranchandramohan replied:
Done.

kparzysz commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -4174,6 +4215,49 @@ struct OmpClauseList {
+
+//    structured-block
+// ...
+struct OmpAssumesPartConstruct {
+  WRAPPER_CLASS_BOILERPLATE(OmpAssumesPartConstruct, Block);
+  CharBlock source;
+};
```
Do we need this class?  Could we just have a `Block` in the construct?

kiranchandramohan replied:
Done.

kparzysz commented on flang/include/flang/Parser/parse-tree.h:
```diff
@@ -4174,6 +4215,49 @@ struct OmpClauseList {
+// assume-construct ->
+//   ASSUME absent-clause | contains-clause | holds-clause | no-openmp-clause |
+//          no-openmp-routines-clause | no-parallelism-clause
+struct OpenMPAssumeConstruct {
+  TUPLE_CLASS_BOILERPLATE(OpenMPAssumeConstruct);
+  std::tuple<Verbatim, OmpClauseList> t;
+  CharBlock source;
+};
```
The `OpenMPDeclarativeAssumes` below is what we need (based on the name, that is).  The `ASSUMES` (with `S` at the end) is the declarative thing, which applies to the entire compilation unit.  There is no `END ASSUMES`.

kiranchandramohan replied:
Done.

Leporacanthicus committed:
Fix ASSUME -> ASSUMES and vice versa

kiranchandramohan committed:
Fixes for rebase

kiranchandramohan committed:
Fix formatting

kiranchandramohan commented on flang/test/Parser/OpenMP/assumption.f90:
```diff
@@ -0,0 +1,59 @@
+! RUN: %flang_fc1  -fopenmp-version=51 -fopenmp -fdebug-unparse-no-sema %s 2>&1 | FileCheck %s
+!!! RUN: %flang_fc1  -fopenmp-version=51 -fopenmp -fdebug-dump-parse-tree-no-sema %s 2>&1 | FileCheck %s --check-prefix="PARSE-TREE"
```
Nit: Is this commented? Removing the extra `!` looks OK now.

kiranchandramohan commented on flang/test/Lower/OpenMP/Todo/declarative-assumes.f90:
```diff
@@ -0,0 +1,8 @@
+! RUN: %not_todo_cmd %flang_fc1 -emit-fir -fopenmp -fopenmp-version=51 -o - %s 2>&1 | FileCheck %s
```
Nit: duplicate test

kiranchandramohan commented on flang/lib/Parser/unparse.cpp:
```diff
@@ -3117,6 +3139,10 @@ class UnparseVisitor {
     return Walk("", list, comma, suffix);
   }
 
+  void Walk(const OmpClauseList &x, const char *sep = " ") {
+    return Walk(" ", x.v, sep);
+  }
```
Nit: Is this required?

Leporacanthicus committed:
Fix test issues

Also rewrite some code to avoid needing a new function.

kiranchandramohan reviewed (approved):
LG. Please wait for @kparzysz.

kparzysz reviewed (approved):
LGTM

kparzysz committed:
Merge branch 'main' into omp-assume
